### PR TITLE
Fix redirect cookie behavior

### PR DESCRIPTION
**Pull request by tomchristie**
Cookies are not being properly handled by redirect responses.
(Specifically, expired cookies in a response, are still included in any redirect request)

### CODE REVIEW

**Comment by tomchristie** [comment]
Needs a test case really.
Linter failures are unrelated. https://travis-ci.org/encode/httpx/jobs/612451968?utm_medium=notification&utm_source=github_status (grumble grumble 🤣)

**Comment by florimondmanca** [comment]
The desired behavior here is that any cookies on a request should be dropped in case of redirect, right?

File: test-requirements.txt [hunk 1]
```diff
@@ -6,7 +6,6 @@ brotlipy==0.7.*
 cryptography
 flake8
 flake8-bugbear
-flake8-comprehensions
```
Comment (StephenBrown2) [inline test-requirements.txt]: Why drop this plugin?

**Comment by tomchristie** [comment]
> The desired behavior here is that any cookies on a request should be dropped in case of redirect, right?

Not quite. It's that whatever the *client* has in it's cookie store is what we should use.

The case where this issue occurs is going to typically be in a logout flow...

1. A logged-in httpx client issues a POST request to the logout URL, including it's current session cookie.
2. Server clears the session cookie (sets an expired cookie on the response) and returns a redirect (eg. back to the homepage/login page).
3. The httpx client makes the homepage redirect request, but because it has cloned the headersof the original request, it erronously includes a cookie on the request, even tho the client cookie store cleared properly when it got the response at (2)

File: test-requirements.txt (hunk 1 above)
Comment (tomchristie) [reply test-requirements.txt]: It's unneccessarily causing breakages...

Eg. expressions such as...

`self.content = b"".join([part async for part in self.stream()])`

Are failing with "C416 Unnecessary list comprehension - rewrite using list()."

**Comment by tomchristie** [comment]
Righty - I'm happy for this to be reviewed now.
The test cases fail without the two-line fixes, and pass after them.

File: tests/client/test_redirects.py [hunk 2]
```diff
@@ -290,3 +290,72 @@ class MockDispatch(AsyncDispatcher):
...
+class MockCookieDispatch(AsyncDispatcher):
+    async def send(
+        self,
+        request: AsyncRequest,
+        verify: VerifyTypes = None,
+        cert: CertTypes = None,
+        timeout: TimeoutTypes = None,
+    ) -> AsyncResponse:
+        if request.url.path == "/":
+            print(request.headers)
```
Comment (florimondmanca) [inline tests/client/test_redirects.py]: Gotcha :)

```suggestion
```

File: test-requirements.txt (hunk 1 above)
Comment (StephenBrown2) [reply test-requirements.txt]: Can be resolved with `# noqa: C416`, as we want to keep the comprehension _there_, but perhaps not other places (non-async comprehensions).

Comment (florimondmanca) [reply test-requirements.txt]: Related: https://github.com/adamchainz/flake8-comprehensions/issues/196

Comment (adamchainz) [reply test-requirements.txt]: Fix released if you want to add it back :)

**Pull request merged**